opentracing/tracecontext.go

```diff
@@ -0,0 +1,82 @@
+package opentracing
+
+// A `TraceContext` is the smallest amount of state needed to describe a span's
+// identity within a larger [potentially distributed] trace. The `TraceContext`
+// is not intended to encode the span's operation name, timing, or log data,
+// but merely any unique identifiers (etc) needed to contextualize it within a
+// larger trace tree.
+//
+// `TraceContext`s are sufficient to propagate the, well, *context* of a
+// particular trace between processes.
+//
+// `TraceContext` also support a simple string map of "trace tags". These trace
+// tags are special in that they are propagated *in-band*, presumably alongside
+// application data. See the documentation for `SetTraceTag()` for more details
+// and some important caveats.
+type TraceContext interface {
+	// Create a child context for this `TraceContext`, and return both that
+	// child's own `TraceContext` as well as any Tags that should be added to
+	// the child's Span.
+	//
+	// The returned `TraceContext` type must be the same as the type of the
+	// `TraceContext` implementation itself.
+	NewChild() (childCtx TraceContext, childSpanTags Tags)
+
+	// Set a tag on this `TraceContext` that also propagates to future
+	// `TraceContext` children per `NewChild()`.
+	//
+	// `SetTraceTag()` enables powerful functionality given a full-stack
+	// opentracing integration (e.g., arbitrary application data from a mobile
+	// app can make it, transparently, all the way into the depths of a storage
+	// system), and with it some powerful costs: use this feature with care.
+	//
+	// IMPORTANT NOTE #1: `SetTraceTag()` will only propagate trace tags to
+	// *future* children of the `TraceContext` (see `NewChild()`) and/or the
+	// `Span` that references it.
+	//
+	// IMPORTANT NOTE #2: Use this thoughtfully and with care. Every key and
+	// value is copied into every local *and remote* child of this
+	// `TraceContext`, and that can add up to a lot of network and cpu
+	// overhead.
+	//
+	// Returns a reference to this `TraceContext` for chaining, etc.
+	SetTraceTag(key, value string) TraceContext
+
+	// Gets the value for a trace tag given its key. Returns the empty string
+	// if the value isn't found in this `TraceContext`.
+	TraceTag(key string) string
+}
+
+// A simple interface to marshal a `TraceContext` to a binary byte array or a
+// string-to-string map.
+type TraceContextMarshaler interface {
+	// Converts the `TraceContext` into marshaled binary data (see
+	// `TraceContextUnmarshaler.UnmarshalTraceContextBinary()`).
+	MarshalTraceContextBinary(tcid TraceContext) []byte
+	// Converts the `TraceContext` into a marshaled string:string map (see
+	// `TraceContextUnmarshaler.UnmarshalTraceContextStringMap()`).
+	MarshalTraceContextStringMap(tcid TraceContext) map[string]string
```

Comment:
I'm really flexible (or at least suggestible) re how best to integrate this with HTTP headers. I tend to regard them as "a mess" and so my personal bias has generally been to cower in fear (c.f. my original proposal to base64 encode the binary context). But if there are people with strong, well-informed opinions, I will let them come up with sensible guidelines and encoding practices for HTTP headers.